src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RazorComponentRenameEndpoint.cs

```diff
@@ -0,0 +1,292 @@
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var originTagHelperBinding = await GetOriginTagHelperBindingAsync(documentSnapshot, codeDocument, request.Position).ConfigureAwait(false);
+            if (originTagHelperBinding is null)
+            {
+                return null;
+            }
+
+            var originTagDescriptor = originTagHelperBinding.Descriptors.First();
+            var originComponentDocumentSnapshot = await await Task.Factory.StartNew(() =>
+            {
+                return _componentSearchEngine.TryLocateComponentAsync(originTagDescriptor).ConfigureAwait(false);
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (originComponentDocumentSnapshot is null)
+            {
+                return null;
+            }
+
+            var newPath = MakeNewPath(documentSnapshot.FilePath, request.NewName);
+            if (File.Exists(newPath))
+            {
+                return null;
+            }
+
+            var documentChanges = new List<WorkspaceEditDocumentChange>();
+            AddFileRenameForComponent(documentChanges, originComponentDocumentSnapshot, newPath);
+            AddEditsForCodeDocument(documentChanges, originTagHelperBinding, request.NewName, request.TextDocument.Uri, codeDocument);
+
+            var projects = await Task.Factory.StartNew(() =>
+            {
+                return _projectSnapshotManager.Projects.ToArray();
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            var documentPaths = new HashSet<string>();
+            foreach (var project in projects)
+            {
+                foreach (var documentPath in project.DocumentFilePaths)
+                {
+                    if (string.Equals(documentPath, documentSnapshot.FilePath, FilePathComparison.Instance))
+                    {
+                        continue;
+                    }
+                    documentPaths.Add(documentPath);
```

Comment:
Instead of collecting paths here can't you just ask the project for the document snapshot? I think you can query that off of the foreground thread IIRC. That way your "AddEditsForCodeDocument" wouldn't need to be async

Reply:
We would still need to asynchronously get the code documents somewhere. Is it worth creating a struct to contain the document snapshot and code document together so I can get them at once and use them synchronously in AddEdits?

Reply:
> We would still need to asynchronously get the code documents somewhere.

Yup, good point, you'd still need to be async.

> Is it worth creating a struct to contain the document snapshot and code document together so I can get them at once and use them synchronously in AddEdits?

Instead of having a struct can't you just change `AddEditsForCodeDocument` to be a DocumentSnapshot accepting API (`AddEditsForDocumentAsync`) and you can just build the list here when iterating the projects?

Reply:
But even if we accept DocumentSnapshots it still requires an asynchronous call to get the RazorCodeDocument? Perhaps I'm misunderstanding so I'll ask you after the standup.

Reply:
Yup, you'd still have an async call there, you just wouldn't have to re-resolve the document inside of that method

Reply:
Sounds good.